lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.cpp

```diff
@@ -0,0 +1,235 @@
+                      Type outputType, xsmm::UnaryKind unaryKind,
+                      Operation *input, Operation *output,
+                      Operation *transposeOp) {
+  std::string dispatchName = "xsmm_unary_dispatch";
+  std::string invokeName = "xsmm_unary_invoke";
+  Location loc = transposeOp->getLoc();
+
+  ModuleOp module = transposeOp->getParentOfType<ModuleOp>();
+  auto dtype =
+      xsmm::utils::getDataType(rewriter, transposeOp->getOperand(0).getType());
+
+  auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+  SmallVector<Value, 10> dispatchOperands;
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(&*functionOp.getBody().op_begin());
+  xsmm::UnaryKindAttr kind =
+      xsmm::UnaryKindAttr::get(rewriter.getContext(), unaryKind);
+  dispatchOperands.push_back(
+      getConstInt(rewriter, dyn_cast<IntegerAttr>(kind).getInt(), 64));
+
+  dispatchOperands.push_back(
+      getConstInt(rewriter, dyn_cast<DataTypeAttr>(dtype).getInt(), 64));
+
+  DenseI64ArrayAttr dims = DenseI64ArrayAttr::get(
+      rewriter.getContext(), ArrayRef<int64_t>{unaryInfo.m, unaryInfo.n,
+                                               unaryInfo.ldi, unaryInfo.ldo});
+  for (auto idx = 0; idx < dims.size(); idx++) {
+    dispatchOperands.push_back(getConstInt(rewriter, dims[idx], 64));
+  }
+  IntegerAttr unaryFlags = dyn_cast<IntegerAttr>(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+  dispatchOperands.push_back(getConstInt(rewriter, unaryFlags.getInt(), 64));
+
+  auto dispatched = xsmm::utils::buildXsmmCall(
+      rewriter, loc, dtype, dispatchOperands,
+      IntegerType::get(rewriter.getContext(), 64), module,
+      SymbolRefAttr::get(transposeOp->getContext(), dispatchName), nullptr);
+  rewriter.setInsertionPoint(transposeOp);
+  SmallVector<Value> operandRange{
+      getConstInt(rewriter, dyn_cast<DataTypeAttr>(dtype).getInt(), 64),
+      dispatched.getResult(0), input->getOperand(0), output->getOperand(1)};
+  auto invokeCall = xsmm::utils::buildXsmmCall(
+      rewriter, loc, dtype, operandRange, TypeRange(), module,
+      SymbolRefAttr::get(transposeOp->getContext(), invokeName), output);
+  return std::make_pair(&*dispatched, &*invokeCall);
+}
+
+static std::pair<Operation *, Operation *>
+convertTransposeOp(PatternRewriter &rewriter, Operation *transposeOp,
+                   Operation *input, Operation *output, Type outputType) {
+  LLVM_DEBUG(llvm::dbgs() << "BuildTransposeOp\n");
+  Value source = input->getOperand(0);
+  VectorType outType = cast<VectorType>(outputType);
+  if (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::TRANSPOSE,
+                                  outType)) {
+    memref::ExpandShapeOp expandShapeOp =
+        dyn_cast<memref::ExpandShapeOp>(source.getDefiningOp());
+    source = expandShapeOp.getSrc();
+    xsmm::UnaryInfo unaryInfo;
+    unaryInfo.m = expandShapeOp.getSrcType().getShape()[0];
```

Comment:
Why isn't this logic in `getVectorUnaryInfo`?

Reply:
I thought since the code was specific to vnni and wasnt gonna be used elsewhere, there was no need to move it to getVectorUnaryInfo. I'll move it there.